### ISSUE TITLE
Analytic view Exceptions on creation / switching of graphs

### ISSUE THREAD

**Issue by aldebaran30701** [constellation-app/constellation#196]
### Prerequisites

* [X] Put an X between the brackets on this line if you have done all of the following:
    * Running the latest version of Constellation
	* Attached the ***Support Package*** via `Help` > `Support Package`
    * Checked the FAQs: https://github.com/constellation-app/constellation/wiki/FAQ
    * Checked that your issue isn't already filed: https://github.com/constellation-app/constellation/issues
    * Checked that there is not already a module that provides the described functionality: https://github.com/constellation-app/constellation/wiki/Catalogue-of-Repositories

### Description

Illegal Argument Exception or Concurrent Modification Exception when switching between graphs with Analytic view open.
```
java.lang.IllegalArgumentException: Parameter 'BetweennessCentralityPlugin.include_connections_in' does not exist
	at au.gov.asd.tac.constellation.pluginframework.parameters.PluginParameters.getParameter(PluginParameters.java:504)
	at au.gov.asd.tac.constellation.pluginframework.parameters.PluginParameters.setBooleanValue(PluginParameters.java:528)
	at au.gov.asd.tac.constellation.views.analyticview.questions.BestConnectsNetworkQuestion.initialiseParameters(BestConnectsNetworkQuestion.java:75)
	at au.gov.asd.tac.constellation.views.analyticview.AnalyticConfigurationPane.lambda$updateSelectablePluginsParameters$28(AnalyticConfigurationPane.java:476)
	at java.lang.Iterable.forEach(Iterable.java:75)
	at au.gov.asd.tac.constellation.views.analyticview.AnalyticConfigurationPane.updateSelectablePluginsParameters(AnalyticConfigurationPane.java:474)
	at au.gov.asd.tac.constellation.views.analyticview.AnalyticViewTopComponent.lambda$null$7(AnalyticViewTopComponent.java:119)
	at au.gov.asd.tac.constellation.functionality.views.ListeningTopComponent.lambda$manualUpdate$16(ListeningTopComponent.java:330)
	at java.util.HashMap.forEach(HashMap.java:1289)
	at au.gov.asd.tac.constellation.functionality.views.ListeningTopComponent.manualUpdate(ListeningTopComponent.java:325)
	at au.gov.asd.tac.constellation.views.analyticview.AnalyticViewTopComponent.handleNewGraph(AnalyticViewTopComponent.java:162)
	at au.gov.asd.tac.constellation.functionality.views.ListeningTopComponent.newActiveGraph(ListeningTopComponent.java:174)
	at au.gov.asd.tac.constellation.graph.node.pluginframework.FrameworkGraphManager.resultChanged(FrameworkGraphManager.java:90)
	at org.openide.util.lookup.SimpleProxyLookup.checkLookup(SimpleProxyLookup.java:114)
	at org.openide.util.lookup.SimpleProxyLookup.lookup(SimpleProxyLookup.java:159)
	at org.netbeans.modules.openide.windows.GlobalActionContextImpl.blickActionMapImpl(GlobalActionContextImpl.java:132)
	at org.netbeans.modules.openide.windows.GlobalActionContextImpl.blickActionMap(GlobalActionContextImpl.java:106)
	at org.netbeans.modules.openide.windows.GlobalActionContextImpl.propertyChange(GlobalActionContextImpl.java:205)
	at java.beans.PropertyChangeSupport.fire(PropertyChangeSupport.java:335)
	at java.beans.PropertyChangeSupport.firePropertyChange(PropertyChangeSupport.java:328)
	at java.beans.PropertyChangeSupport.firePropertyChange(PropertyChangeSupport.java:263)
	at java.awt.KeyboardFocusManager.firePropertyChange(KeyboardFocusManager.java:1493)
	at java.awt.KeyboardFocusManager.setGlobalPermanentFocusOwner(KeyboardFocusManager.java:780)
	at java.awt.DefaultKeyboardFocusManager.dispatchEvent(DefaultKeyboardFocusManager.java:615)
	at java.awt.Component.dispatchEventImpl(Component.java:4760)
	at java.awt.Container.dispatchEventImpl(Container.java:2297)
	at java.awt.Component.dispatchEvent(Component.java:4711)
	at java.awt.EventQueue.dispatchEventImpl(EventQueue.java:760)
	at java.awt.EventQueue.access$500(EventQueue.java:97)
	at java.awt.EventQueue$3.run(EventQueue.java:709)
	at java.awt.EventQueue$3.run(EventQueue.java:703)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:74)
	at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:84)
	at java.awt.EventQueue$4.run(EventQueue.java:733)
	at java.awt.EventQueue$4.run(EventQueue.java:731)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:74)
	at java.awt.EventQueue.dispatchEvent(EventQueue.java:730)
	at org.netbeans.core.TimableEventQueue.dispatchEvent(TimableEventQueue.java:159)
[catch] at java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:205)
	at java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:116)
	at java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:105)
	at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
	at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:93)
	at java.awt.EventDispatchThread.run(EventDispatchThread.java:82)
```
```
java.util.ConcurrentModificationException
	at java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1388)
	at java.util.stream.ReferencePipeline$Head.forEach(ReferencePipeline.java:647)
	at au.gov.asd.tac.constellation.pluginframework.parameters.PluginParameter.fireChangeEvent(PluginParameter.java:586)
	at au.gov.asd.tac.constellation.pluginframework.parameters.PluginParameter.setObjectValue(PluginParameter.java:475)
	at au.gov.asd.tac.constellation.pluginframework.parameters.PluginParameters.lambda$updateParameterValues$2(PluginParameters.java:304)
	at java.util.Collections$UnmodifiableMap$UnmodifiableEntrySet.lambda$entryConsumer$0(Collections.java:1577)
	at java.util.LinkedHashMap$LinkedEntrySet.forEach(LinkedHashMap.java:671)
	at java.util.Collections$UnmodifiableMap$UnmodifiableEntrySet.forEach(Collections.java:1582)
	at au.gov.asd.tac.constellation.pluginframework.parameters.PluginParameters.updateParameterValues(PluginParameters.java:302)
	at au.gov.asd.tac.constellation.views.analyticview.AnalyticConfigurationPane.lambda$updateSelectablePluginsParameters$28(AnalyticConfigurationPane.java:475)
	at java.lang.Iterable.forEach(Iterable.java:75)
	at au.gov.asd.tac.constellation.views.analyticview.AnalyticConfigurationPane.updateSelectablePluginsParameters(AnalyticConfigurationPane.java:474)
	at au.gov.asd.tac.constellation.views.analyticview.AnalyticViewTopComponent.lambda$null$7(AnalyticViewTopComponent.java:119)
	at au.gov.asd.tac.constellation.functionality.views.ListeningTopComponent.lambda$manualUpdate$16(ListeningTopComponent.java:330)
	at java.util.HashMap.forEach(HashMap.java:1289)
	at au.gov.asd.tac.constellation.functionality.views.ListeningTopComponent.manualUpdate(ListeningTopComponent.java:325)
	at au.gov.asd.tac.constellation.views.analyticview.AnalyticViewTopComponent.handleNewGraph(AnalyticViewTopComponent.java:162)
	at au.gov.asd.tac.constellation.functionality.views.ListeningTopComponent.newActiveGraph(ListeningTopComponent.java:174)
	at au.gov.asd.tac.constellation.graph.node.pluginframework.FrameworkGraphManager.resultChanged(FrameworkGraphManager.java:90)
	at org.openide.util.lookup.SimpleProxyLookup.checkLookup(SimpleProxyLookup.java:114)
	at org.openide.util.lookup.SimpleProxyLookup.lookup(SimpleProxyLookup.java:159)
	at org.netbeans.modules.openide.windows.GlobalActionContextImpl.blickActionMapImpl(GlobalActionContextImpl.java:132)
	at org.netbeans.modules.openide.windows.GlobalActionContextImpl.blickActionMap(GlobalActionContextImpl.java:106)
	at org.netbeans.modules.openide.windows.GlobalActionContextImpl.propertyChange(GlobalActionContextImpl.java:205)
	at java.beans.PropertyChangeSupport.fire(PropertyChangeSupport.java:335)
	at java.beans.PropertyChangeSupport.firePropertyChange(PropertyChangeSupport.java:328)
	at java.beans.PropertyChangeSupport.firePropertyChange(PropertyChangeSupport.java:263)
	at java.awt.KeyboardFocusManager.firePropertyChange(KeyboardFocusManager.java:1493)
	at java.awt.KeyboardFocusManager.setGlobalPermanentFocusOwner(KeyboardFocusManager.java:780)
	at java.awt.DefaultKeyboardFocusManager.dispatchEvent(DefaultKeyboardFocusManager.java:615)
	at java.awt.Component.dispatchEventImpl(Component.java:4760)
	at java.awt.Container.dispatchEventImpl(Container.java:2297)
	at java.awt.Component.dispatchEvent(Component.java:4711)
	at java.awt.EventQueue.dispatchEventImpl(EventQueue.java:760)
	at java.awt.EventQueue.access$500(EventQueue.java:97)
	at java.awt.EventQueue$3.run(EventQueue.java:709)
	at java.awt.EventQueue$3.run(EventQueue.java:703)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:74)
	at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:84)
	at java.awt.EventQueue$4.run(EventQueue.java:733)
	at java.awt.EventQueue$4.run(EventQueue.java:731)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:74)
	at java.awt.EventQueue.dispatchEvent(EventQueue.java:730)
	at org.netbeans.core.TimableEventQueue.dispatchEvent(TimableEventQueue.java:159)
[catch] at java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:205)
	at java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:116)
	at java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:105)
	at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
	at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:93)
	at java.awt.EventDispatchThread.run(EventDispatchThread.java:82)

```
### Steps to Reproduce

1. Open Constellation
2. Open New Analytic Graph
3. Perform Sphere Graph Builder (anything to have a non - empty graph)
4. Save Graph (Unsure if completely necessary)
5. Open Analytic View
6. Expand Categories
7. Check Between-ness Centrality Analytic
8. Open New Graph
9. Exception occurs

**Expected behaviour:** 
No error message

**Actual behaviour:** 
Error messages / Exceptions

**Reproduces how often:** 
50-60%

### Additional Information
If analytic view was already opened, that may effect the replication.

**Comment by aldebaran30701** [comment]
After step 7, if you **run the plugin** -> _close_ **Analytic View** -> _open_ **Analytic View** you will see the following exception:
```
java.util.ConcurrentModificationException
	at java.util.ArrayList.forEach(ArrayList.java:1260)
	at java.util.Collections$UnmodifiableCollection.forEach(Collections.java:1082)
	at au.gov.asd.tac.constellation.pluginframework.gui.MultiChoiceInputPane.lambda$null$4(MultiChoiceInputPane.java:103)
	at com.sun.javafx.application.PlatformImpl.lambda$null$5(PlatformImpl.java:295)
	at java.security.AccessController.doPrivileged(Native Method)
	at com.sun.javafx.application.PlatformImpl.lambda$runLater$6(PlatformImpl.java:294)
[catch] at com.sun.glass.ui.InvokeLaterDispatcher$Future.run(InvokeLaterDispatcher.java:95)
	at com.sun.glass.ui.win.WinApplication._runLoop(Native Method)
	at com.sun.glass.ui.win.WinApplication.lambda$null$3(WinApplication.java:177)
	at java.lang.Thread.run(Thread.java:748)
```

**Comment by aldebaran30701** [comment]
I've had a further look with @formalhaut69 regarding this most recent comment. 
Initially we got concurrent modification errors at MultiChoiceInputPane.java:103
Replication was few and far between. An attempt at changing the iterator style was used, it seemed to stop the error at line 103.
Upon further testing we had the same error pop up at line MultiChoiceInputPane.java:99 on 
list.addAll(); 
Some digging proved that the error may also lead to something on another thread editing the variables during the same time. This leads us to believe that there is a possibility of another thread iterating the collection at the same time as this section of code runs on a UI thread using Platform.runLater().

At risk of diving into something out of our depth, would you have any suggestions or insights into this issue. @arcturus2

**Comment by arcturus2** [comment]
Hey @aldebaran30701 and @formalhaut69 so the issue will most likely be in the Analytic view code doing the wrong thing by reading and writing to the the parameter from different threads . I'm suggesting looking at the Analytic view code first because the Core Plugin Parameters code like the `MultiChoiceInputPane` etc has been around for longer and gets used by other views like the Data Access View without a problem.

`Platform.runLater()` pushes the work to the JavaFX thread which is single threaded (as far as I know) so my bet is that during a call to it from inside `Platform.runLater()` the parameters are being called directly too (outside JavaFX).

We'd need to first find out what object is being read and written to first so that is the hard part.

Something I like to do is add an `assert` statement to confirm if you are or aren't on the graphics thread depending on the question your trying to answer. This helped me track down some issues with the `MapViewTileRenderer.java` which had performance and ConcurrentModExceptions with objects being called and changed from many threads.

Perhaps if I explain how I tackled that, it might give you ideas on how to solve this. I'll also have a go at trying to fix this one.

What I did was add `assert !SwingUtilities.isEventDispatchThread();` to make sure process intensive methods were not running on the graphics thread. If you have a look at `MapViewTileRenderer.java` 
you will see that they are still there and I left them so that future changes didn't re-introduce the performance problem. Given the Analytic View is JavaFX the assert would be `assert !Platform.isFxApplicationThread()`.

It turned out in that example it was the `map` object and the way to make it thread "safe" was to synchronise the call using a "LOCK" Object.

The lock Object was defined by doing something like:

```java
public static final Object LOCK = new Object();
```

I then wrapped any call to the `map` object with a synchronise call like:

```java
    synchronized (LOCK) {
        map.addMarker(marker);
    }
```

How does this work? Well it's a novel idea I learnt from a java guru and the idea is that when you call `synchronized` from multiple thread's it forces the other threads to block. You need to synchronise on something so I'm following this convention of synchronising on a `LOCK` Object.

Is any of that useful or make sense?

**Comment by formalhaut69** [comment]
I've made a pull request to fix this issue but the second concurrent modification exception that @aldebaran30701 mentions in his comment relating to the MultiChoiceInputPane appears to be a completely different issue. I think we should close this one and open up separate ticket for that one. What do you think @arcturus2?

**Comment by arcturus2** [comment]
> I've made a pull request to fix this issue but the second concurrent modification exception that @aldebaran30701 mentions in his comment relating to the MultiChoiceInputPane appears to be a completely different issue. I think we should close this one and open up separate ticket for that one. What do you think @arcturus2?

OK @formalhaut69, happy to close this and make a separate ticket for the that exception...the only thing is that I'm not able to reproduce it with the latest master. If you or @aldebaran30701 have time and can reproduce it then please create a ticket for me.
